pkg/obfuscate/appsec.go

```diff
@@ -0,0 +1,538 @@
+		return val
+	}
+	return output
+}
+
+type appsecEventsObfuscator struct {
+	keyRE, valueRE *regexp.Regexp
+}
+
+type unexpectedScannerOpError int
+
+func (err unexpectedScannerOpError) Error() string {
+	return fmt.Sprintf("unexpected json scanner operation %d", err)
+}
+
+// inputDiff holds the list of diffs, built by the obfuscator, and to ultimately
+// apply to the input.
+type inputDiff []struct {
+	from, to int
+	value    string
+}
+
+// Add a new entry to the list. Note that due to how the apply() method works
+// by creating the output string by walking this list, the list entries must be
+// sorted in ascending "input offset" order.
+func (d *inputDiff) add(from, to int, value string) {
+	elt := struct {
+		from  int
+		to    int
+		value string
+	}{
+		from:  from,
+		to:    to,
+		value: value,
+	}
+	diff := *d
+	l := len(diff)
+	i := sort.Search(l, func(i int) bool {
+		return (*d)[i].to > from
+	})
+	if i == l {
+		*d = append(diff, elt)
+		return
+	}
+	*d = append(diff[:i+1], diff[i:]...)
+	(*d)[i] = elt
+}
+
+// Apply the diff to the input if any.
+func (d inputDiff) apply(input string) string {
+	from := 0
+	var output strings.Builder
+	for _, diff := range d {
+		output.WriteString(input[from:diff.from])
+		from = diff.to + 1
+		output.WriteString(diff.value)
+	}
+	output.WriteString(input[from:])
+	return output.String()
+}
+
+// Append the given diff to the current one, shifted by the given offset.
+// Cf. method obfuscateRuleMatchParameter()
+func (d *inputDiff) appendTo(diff *inputDiff, offset int) {
+	for _, d := range *d {
+		diff.add(d.from+offset, d.to+offset, d.value)
+	}
+}
+
+// Entrypoint of the appsec obfuscator based on a json scanner and which tries
+// to only obfuscate the appsec rule parameters of the given input by searching
+// and targeting objects of the form
+// `{ "parameters": <appsec rule match parameters>, <any other keys>: <any other values> }`
+// When such object is found, its parameters' value is obfuscated.
+func (o *appsecEventsObfuscator) obfuscate(input string) (output string, err error) {
+	var (
+		scanner scanner
+		diff    inputDiff
+	)
+	scanner.reset()
+	keyFrom := -1
+	keyTo := -1
+	// Scan the input to discover its objects and see if they have the
+	// `parameters` key we are looking for.
+	for i := 0; i < len(input); i++ {
+		switch scanner.step(&scanner, input[i]) {
+		case scanError:
+			return input, scanner.err
+
+		case scanBeginLiteral:
+			// Possibly the beginning of an object key
+			if input[i] == '"' {
+				keyFrom = i
+			}
+		case scanContinue:
+			// Continue the literal value
+			if keyFrom != -1 && input[i] == '"' && input[i-1] != '\\' {
+				// Ending double-quote found
+				keyTo = i
+				// Only scanObjetKey will confirm this was an object key
+			}
+		case scanObjectKey:
+			// Object key scanned and keyFrom and keyTo were set according to
+			// the previous scanBeginLiteral and scanContinue scanner operations
+			if input[keyFrom:keyTo+1] == `"parameters"` {
```

Comment:
Shouldn't we break the outer loop here? Why are we continuing? We found `"parameters"`, we're done, right? Are there going to be other `"parameters"` anywhere deeper in the JSON that we care about?

Reply:
Yes, the overall payload of appsec events is bigger than that and corresponds to the `jq` query `.triggers[].rule_matches[].parameters[]`, so there can be other parameters following in `input`.
This function is a simplification of this jq query where I simplified it to the minimum by just looking for all the objects having this `"parameters"` key rather than having to handle the full logic of scanning the entire payload.

I basically want to try to avoid agent modifications for each appsec event payload modification that doesn't impact the sensitive data obfuscation (like adding a fields, changing the enveloppe, etc.).

Reply:
Got it. Thanks for explaining.